### ISSUE TITLE
How to run multiple tasks sync/async?

### ISSUE THREAD

**Issue by JoshMcCullough** [OverZealous/run-sequence#37]
Say I have these tasks: clean, compile:ts, compile:scss, min:js, min:css, copy:bower

The should be run in this order:

- `clean` runs first.
- Then, `compile:ts` and `min:js` run in sequence while `compile:scss` and `min:css` run in sequence.
- Then, once all of that is done, `copy:bower` is run.

So basically, clean, do JS-stuff while doing CSS-stuff, then finally do bower stuff.

Couldn't see how to do this using run-sequence, is it possible?

**Comment by OverZealous** [comment]
You can only have one set of sequenced items per `run-sequence`, but each step can have multiple parallel steps.  See [Usage](https://github.com/OverZealous/run-sequence#usage).

You can also use sub tasks to group run-sequence steps.  Honestly, though, I'd just run `('clean', ['compile:ts', 'min:js'], ['comile:scss', 'min:css'], 'copy:bower')`, and not worry about running JS parallel to the styles, since you won't gain much.  Or even better, use gulp's native task-dependencies to handle some of this.  There's no reason you need to use run-sequence for standard dependencies, only steps that aren't always dependent.

In your example above, you shouldn't even be running a separate `min` step, you should be piping the output from your compile process straight into the minification, so you don't have to hit the disk at all.  You can use simple functions, or something like [lazypipe](https://github.com/OverZealous/lazypipe), to reuse build steps for both development and production.  If you do this, then you can do something like `('clean', ['js-prod', 'css-prod'], 'copy:bower')`

**Comment by JoshMcCullough** [comment]
Thanks for responding, appreciate it. But I disagree that, for example, `min:js` is dependent on `compile:ts`. Yes, at some point it is, but not always. And I get the sense that that is why this package exists.

In my case, which I think would be very normal, `compile:ts` happens whenever some TS file changes. And `min:js` happens during build/test/deployment (not with every change to TS files). In theory, the JS files are up to date by the time build/test/deployment happens, so minification can continue w/out depending on compilation (it's already done). _In theory._ In practice, things are rebuilt anyway and you're point is correct that I should just setup the dependency.

I was asking more philosophically, I suppose.

I'll continue w/the dependency route for now! Appreciate your time.

**Comment by OverZealous** [comment]
I didn't mean as a dependency—I mean there shouldn't be separate tasks at all.  A separate task implies that something was written to disk and is being read back in—that's not really the gulp way.

Rather, you create a single task which passes code directly from compile to minification, but only for production.  To reuse code, you don't wrap it in tasks, you wrap it in functions.

For example, you might have a function like:

```js
function compileTS() {
    return gulp.src(...).pipe(...);
}

gulp.task('compile-dev', function() {
    return compileTS().pipe(gulp.dest(...));
});
gulp.task('compile-prod', function() {
    return compileTS().pipe(minificationFunction()).pipe(gulp.dest(...));
});
```

You can even pipe to `gulp.dest()` twice in the prod function if you want both the minified and unminified files, since it's a pass-through.

**Comment by OverZealous** [comment]
Also: lazypipe (linked above) makes it even easier to build up reusable pipe streams.

**Comment by JoshMcCullough** [comment]
Makes sense, thanks.
